### PR TITLE
feat: expose per-page direction and confidence signals

### PR DESCRIPTION
**Pull request by SomSamantray**
## Summary

Callers can now self-diagnose OCR-routing decisions locally. Per-page extraction results and Full-mode process results expose the reading direction (`ltr` / `rtl` / `mixed`) and a per-page text-confidence score (0.0-1.0) on all four surfaces (Rust, Python, napi, wasm). The machine-readable OCR-reason vocabulary is documented and typed as a union in the hand-maintained wasm TS declarations.

Previously a wrong routing decision was invisible until someone diffed against another extractor: an RTL misfire or an over-flagged page looked identical to a correct result. These signals make both self-diagnosing.

## What changed

- `PageMarkdown.direction` + `PageMarkdown.confidence` on per-page extraction results (Rust / Python / napi).
- `PdfProcessResult.page_signals` (per-page direction + confidence, 1-indexed) on Full-mode process results across all four surfaces, including wasm.
- Confidence formula (documented in `docs/rust-api.md`): 1.0 clean text; 0.5 at the 500-bps replacement-density OCR threshold; 0.15 binary garbled evidence (substitution cipher, PUA/C1 spans); 0.0 no usable text.
- Direction semantics: per-line RTL/LTR counts mirroring `is_rtl_text`, computed over the same items that produce the emitted markdown (page-number footers removed first); `mixed` means RTL ordering applied on some lines, so callers key on `direction != "ltr"`.
- OCR-reason vocabulary documented (4 literals) with multi-value and merge-order semantics; `phantom_empty_row`/`detection_error`/`multi_row_in_cell` explicitly excluded as TSR table-fallback labels.
- No behavior changes: extracted text, markdown, `pdf_type`, document-level confidence, classification, and CLI JSON output are byte-identical (existing snapshots unchanged).

## Design decisions

- Per-page confidence lives on the extraction path and Full-mode process results (where every processed page is analyzed); the lightweight classify path keeps document-level confidence because detection samples pages (default `Sample(8)`).
- The new public fields are compile-time breaking for Rust struct-literal construction and napi-generated `index.d.ts` consumers (semver-minor); runtime output is unchanged.
- Documented divergences: extract vs Full-mode confidence bases differ on garbage-stripped pages; `0.15` is inexact after f32->f64 widening on Python/napi/wasm (compare with `abs(diff) < 1e-3`).

## Validation

- `cargo test`: 996 passing (832 unit + 162 integration + 2 doc), zero snapshot drift; `cargo clippy -- -D warnings` and `cargo fmt --check` clean; `cargo build --release` clean.
- wasm: `cargo check --target wasm32-unknown-unknown` clean; `wasm-pack test` runs in CI (not available locally).
- Python: `maturin develop` + pytest 65/66 (the `encrypted-secret123` fixture failure is pre-existing on `main` - the test calls `process_pdf` without a password; verified identical on the base commit).
- napi: `npm run build` + `node test.mjs` all pass.
- New tests cover the confidence anchors (0.15/0.5/0.0/1.0) through the public API, RTL/mixed/blank/out-of-range page signals, Analyze mode, page-filter edge cases, and the KTD2 footer-removal semantics.

Related: #217

Plan: `docs/plans/2026-08-06-001-feat-expose-detector-signals-plan.md`

---

[Compound Engineering](https://github.com/EveryInc/compound-engineering-plugin)

---
## Summary by cubic
Expose per-page reading direction and text-confidence across Rust, Python, `napi`, and `wasm` to help diagnose OCR routing, with stricter 0.0 grading when only non-text items remain and small binding fixes.

- **New Features**
  - `PageMarkdown.direction` and `PageMarkdown.confidence` (0-indexed).
  - `PdfProcessResult.page_signals` (1-indexed) in Full/Analyze; empty in DetectOnly and Scanned/ImageBased.
  - Direction computed over markdown-emitted items after page-number removal; `mixed` means RTL applied on some lines.
  - Confidence anchors: 1.0 clean, 0.5 at 500 bps, 0.15 binary garble, 0.0 when no text remains; OCR reasons documented and typed in `wasm` (`OcrReason` union).

- **Bug Fixes**
  - Confidence is 0.0 when no text survives page-number removal, including pages where only non-text items remain (image-only pages default to `ltr`/0.0).
  - `page_signals` cleared when a Mixed PDF is upgraded to Scanned.
  - Form fields included in direction classification.
  - Python: `PageSignals` registered as a module class.

<sup>Written for commit 3fa22c77d6e8fef96955b8ad413c2fe9b5a6f7c1. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/firecrawl/pdf-inspector/pull/277?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: src/python.rs [hunk 1]
```diff
@@ -85,6 +89,31 @@ impl PyPageOcrReasons {
     }
 }
 
+/// Per-page reading direction and text-confidence for Full-mode results.
+#[pyclass(name = "PageSignals")]
```
[the remark this answers is not in the document]
Comment (SomSamantray) [reply src/python.rs:93]: Fixed: PyPageSignals is now registered via m.add_class::<PyPageSignals>()? in pdf_inspector(), so pdf_inspector.PageSignals is importable. Verified with a runtime hasattr check.

File: src/text_utils.rs [hunk 2]
```diff
@@ -144,6 +156,70 @@ pub(crate) fn sort_line_items(items: &mut [TextItem]) {
...
+/// and none is LTR-dominant, `Mixed` when both occur, `Ltr` otherwise. The
+/// grouping is an approximation: the extractor additionally splits same-baseline
+/// items by column/x-position, which this helper does not model.
+pub(crate) fn page_direction<'a, I>(items: I) -> PageDirection
+where
+    I: IntoIterator<Item = &'a TextItem>,
+{
+    let mut items: Vec<&TextItem> = items
+        .into_iter()
+        .filter(|item| matches!(item.item_type, ItemType::Text))
```
[the remark this answers is not in the document]
Comment (SomSamantray) [reply src/text_utils.rs]: Fixed: page_direction now includes ItemType::FormField in its text-item filter, so form-field text participates in the direction signal like it does in direction-aware layout.

File: src/lib.rs [hunk 3]
```diff
@@ -3727,6 +3800,7 @@ fn process_document(
         gid_pages,
         text_quality_pages,
         text_quality_reasons_by_page,
+        page_signals,
```
[the remark this answers is not in the document]
Comment (SomSamantray) [reply src/lib.rs:3935]: Fixed: page_signals is cleared when a Mixed PDF's extracted markdown is upgraded to Scanned, so the documented 'empty for Scanned/ImageBased' invariant holds for upgraded results too.

File: src/lib.rs [hunk 4]
```diff
@@ -3986,6 +4102,23 @@ fn suspected_garbled_reason() -> String {
...
+pub(crate) fn page_signal<'a, I>(
+    page_items: I,
+    confidence_by_page: &BTreeMap<u32, f32>,
+    page_1idx: u32,
+) -> (PageDirection, f32)
+where
+    I: IntoIterator<Item = &'a TextItem>,
+{
+    let direction = crate::text_utils::page_direction(page_items);
+    let confidence = confidence_by_page.get(&page_1idx).copied().unwrap_or(0.0);
```
[the remark this answers is not in the document]
Comment (SomSamantray) [reply src/lib.rs:4270]: Fixed: page_signal now grades 0.0 (and ltr) when no text items survive page-number removal, so a page whose only content was a removed folio no longer reports confidence 1.0.

File: src/text_quality.rs [hunk 5]
```diff
@@ -297,9 +310,26 @@ pub(crate) fn analyze_text_quality(items: &[TextItem]) -> TextQualityReport {
...
+/// Per-page text-confidence (0.0–1.0) per the confidence contract:
+/// - no extractable text → 0.0
+/// - binary garbled evidence (cipher-garble or a Strong-span flag) → 0.15
+/// - otherwise density-graded by replacement chars per 10,000 non-whitespace
+///   chars: 1.0 clean, 0.5 at 500 bps, 0.0 at ≥ 1000 bps.
+fn page_confidence(evidence: &PageTextQualityEvidence) -> f32 {
+    if evidence.chars == 0 {
+        return 0.0;
+    }
+    if evidence.cipher_garble.looks_garbled() || evidence.strong_flag {
```
[the remark this answers is not in the document]
Comment (SomSamantray) [reply src/text_quality.rs:326]: Declined with reasoning: the cipher-garble heuristic (looks_garbled) is pre-existing, gated (ascii_letters >= 200, vowel-ratio, permuted-language shape checks), and validated against the eval corpus with zero false positives; this PR only surfaces its existing result as the 0.15 confidence band. Changing the heuristic would alter existing needs_ocr behavior, which this PR's behavior-neutrality contract (no change to OCR routing) explicitly forbids. The 0.15 band and its divergence caveats are documented in docs/rust-api.md.

File: src/text_utils.rs [hunk 6]
```diff
@@ -144,6 +156,70 @@ pub(crate) fn sort_line_items(items: &mut [TextItem]) {
...
+        .collect();
+    if items.is_empty() {
+        return PageDirection::Ltr;
+    }
+    items.sort_by(|a, b| a.y.total_cmp(&b.y));
+
+    // Group into lines by Y with the markdown line-grouping tolerance (items
+    // are sorted by Y, so clusters are contiguous).
+    const Y_TOLERANCE: f32 = 3.0;
```
[the remark this answers is not in the document]
Comment (SomSamantray) [reply src/text_utils.rs:185]: Addressed by documentation: the approximation (3pt Y-tolerance, no column/x split) is now stated explicitly in the public direction contract in docs/rust-api.md and in the helper doc. The extractor's exact line grouping is column/x-aware; replicating it inside a per-page diagnostic signal would duplicate that machinery. The signal's safe guidance (test direction != "ltr") is unaffected.

File: napi/README.md [hunk 7]
```diff
@@ -126,3 +162,14 @@ Prebuilt binaries ship as platform-specific packages installed automatically via
 ## License
 
 MIT
+
+### OCR reason vocabulary
```
[the remark this answers is not in the document]
Comment (SomSamantray) [reply napi/README.md]: Fixed: the OCR reason vocabulary section now sits under Types (before Platforms), where readers find it with the type definitions it explains.

File: docs/python.md [hunk 8]
```diff
@@ -113,13 +113,19 @@ class PdfResult:                     # process_pdf / detect_pdf
...
     ocr_reasons_by_page: list[PageOcrReasons]
+    page_signals: list[PageSignals]  # per-page direction + confidence (present when extraction ran)
     title: str | None
     confidence: float                # 0.0 - 1.0
     is_complex_layout: bool
     pages_with_tables: list[int]
     pages_with_columns: list[int]
     has_encoding_issues: bool        # broken font encodings — consider OCR fallback
 
+class PageSignals:                   # per-page direction + text-confidence (process_pdf Full mode)
```
[the remark this answers is not in the document]
Comment (SomSamantray) [reply docs/python.md]: Fixed: the PageSignals doc block now says the list is present whenever extraction ran (Full or Analyze), matching the PdfResult.page_signals contract.

File: docs/plans/2026-08-06-001-feat-expose-detector-signals-plan.md [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+- **Dependencies:** U1, U6
+- **Files:**
+  - `napi/src/lib.rs` — `PageMarkdownResult` struct and the `extract_pages_markdown` mapping (~line 637); `PdfResult` gains `page_signals` (a `Vec<PageSignals>` `#[napi(object)]` struct), populated in `to_napi_result` (~line 130)
+  - `napi/test.mjs` — assertions on the new fields
+  - `napi/README.md` — per-page types and process-result documentation
+- **Approach:**
+  1. Add `direction: String` and `confidence: f64` to the `#[napi(object)]` `PageMarkdownResult` (napi auto-converts to camelCase; both names are already single words). The core `f32` value is widened to `f64`.
+  2. Add a `PageSignals` `#[napi(object)]` struct (`page: u32`, `direction: String`, `confidence: f64`) and `PdfResult.page_signals`, populated in `to_napi_result` (empty when extraction did not run).
+  3. Populate them in the `extract_pages_markdown` and `process_pdf` mappings (mirroring the existing `needs_ocr`/`ocr_reason` mapping).
+  4. Assert the fields in `napi/test.mjs` (field presence + expected anchor values 0.0, 0.15, 0.5, 1.0 on the shared fixtures).
```
[the remark this answers is not in the document]
Comment (SomSamantray) [reply docs/plans/2026-08-06-001-feat-expose-detector-signals-plan.md]: Fixed: the U4 test-step wording now specifies exact comparison for 0.0/0.5/1.0 and tolerance-based comparison for 0.15 on the f64 surfaces, consistent with the KTD4 widening note.

File: tests/integration_tests.rs [hunk 10]
```diff
@@ -1600,6 +1601,29 @@ fn test_extract_regions_mem_identity_h_needs_ocr() {
 /// and the embedded subset font has no `cmap` table to recover from. The
 /// resulting ciphertext is 100% printable ASCII, so it must be caught by the
 /// substitution-cipher statistics and routed to OCR instead of served silently.
+#[test]
```
[the remark this answers is not in the document]
Comment (SomSamantray) [reply tests/integration_tests.rs:1604]: Fixed: the ParseBench doc comment is restored to test_extract_pages_mem_shifted_cipher_tounicode_needs_ocr (including its re-added confidence assert), and the new tests sit after it with a single #[test] each.

**Comment by SomSamantray** [comment]
## Babysit run report

Autonomous review/CI pass complete. All 10 cubic review threads resolved (8 findings fixed and verified, 1 addressed by documentation, 1 declined with reasoning on the pre-existing cipher heuristic). Branch updated to contain `main` (merge of #231 resolved; GitHub mergeability cache pending recompute — the branch contains the base commit, so this is stale cache, not a real conflict).

### Residuals

- **CI workflow not yet dispatched.** The `ci.yml` `pull_request` workflow has not created a run for this fork PR after \~25 minutes. This repo gates some fork-PR runs on maintainer approval (cf. the `action_required` run on PR #270), so the run may appear as pending-approval and need a maintainer to trigger it. Nothing on the PR side can unblock this.
- **Mergeability state stale.** GitHub reports `dirty` although the branch contains the base commit (`git merge-base --is-ancestor origin/main HEAD` passes); a merge would be clean. Pending GitHub recompute.

Local verification: 998 Rust tests passing (incl. upstream #231's tests), clippy/fmt clean, python 68/69 (pre-existing encrypted-fixture failure), napi all pass, wasm32 check clean.

File: src/lib.rs [hunk 11]
```diff
@@ -4016,6 +4141,28 @@ fn suspected_garbled_reason() -> String {
...
+pub(crate) fn page_signal<'a, I>(
+    page_items: I,
+    confidence_by_page: &BTreeMap<u32, f32>,
+    page_1idx: u32,
+) -> (PageDirection, f32)
+where
+    I: IntoIterator<Item = &'a TextItem>,
+{
+    let items: Vec<&TextItem> = page_items.into_iter().collect();
+    if items.is_empty() {
```
[the remark this answers is not in the document]
Comment (SomSamantray) [reply src/lib.rs]: Fixed: page_signal now requires at least one Text/FormField item to survive page-number removal before grading. An image-only page (or a page whose only text was a removed folio) reports ltr/0.0 instead of inheriting the removed text's confidence. Unit tests added (page_signal_grades_zero_when_only_non_text_remains, page_signal_grades_form_field_text).

File: tests/integration_tests.rs [hunk 12]
```diff
@@ -9,8 +9,8 @@ use pdf_inspector::{
     extract_pages_markdown_mem, extract_tables_in_regions_mem, extract_text,
```
[the remark this answers is not in the document]
Comment (SomSamantray) [reply tests/integration_tests.rs:9]: Fixed: the #[test] attribute on test_extract_regions_mem_multiple_regions_per_page is restored (it was lost during the earlier test relocation), so the multi-regions coverage runs again (integration suite back to 166 passing).

**Comment by princeamir56** [comment]
# PR Review — #277: feat: expose per-page direction and confidence signals

**Repository**: firecrawl/pdf-inspector
**Author**: @SomSamantray
**Branch**: feat/expose-detector-signals → main
**Date**: 2026-08-07
**Reviewed by**: PR Review Agent (Orchestrator + 3 agents)

---

## 📋 Summary

**What changed:** feat: expose per-page direction and confidence signals. Touches 17 files (root, docs/, napi/, src/, …) with +1423/-12 lines.

**Why it exists:** `bug fix` — title/commits reference a fix or bug.

**Impact areas:** `README.md`, `docs/plans`, `docs/python.md`, `docs/residual-review-findings`, `docs/rust-api.md`, `napi/README.md`

**Complexity score:** High — 17 files, 1435 changed lines, 6 impact areas.

**Files changed summary:**
- `docs/plans/2026-08-06-001-feat-expose-detector-signals-plan.md` — new file (md) (+285/-0)
- `tests/integration_tests.rs` — modified (rs) (+257/-2)
- `src/text_utils.rs` — modified (rs) (+182/-2)
- `src/lib.rs` — modified (rs) (+176/-0)
- `src/text_quality.rs` — modified (rs) (+146/-0)

---

## 🔒 Security Analysis

**Risk level:** 🟢 Clean

✅ No security issues detected

**Categories checked:** secrets ✅ · injection ✅ · auth ✅ · dependencies ✅ · hardcoded ✅ · input-validation ✅ · sql ✅ · xss ✅ · cors/csp ✅ · path-traversal ✅ · ssrf ✅ · deserialization ✅ · crypto ✅ · open-redirect ✅ · cookie ✅ · tls ✅ · prototype-pollution ✅ · header-injection ✅

**External scanners:** gitleaks, Semgrep OSS, Trivy ran — no findings beyond the rule engine.

_Suppress a false positive by adding `// pr-agent-ignore` or `// nosec` on the flagged line._

---

## 📚 Documentation

**Missing docstrings/JSDoc:**
- `tests/test_python.py` → `test_page_signals_present_on_detect_only_result()`

**README impact:** No README update required.

**Changelog entry:**
```markdown
## [Unreleased]
### Added
- feat: expose per-page direction and confidence signals.
```

**Auto-generated docstrings:**
```python
def test_page_signals_present_on_detect_only_result(...):
    """One-line summary of what the function does.



    Returns:
        Description of the return value.
    """
```

---

## ✅ Review Decision

| Agent | Status | Risk |
|---|---|---|
| Summary | ✅ Complete | — |
| Security | 🟢 Clean | Clean |
| Documentation | 🟡 Updates needed | Low |

**Overall recommendation**: NEEDS DISCUSSION

---
*Generated by PR-Review-Agent on 2026-08-07 01:35:24 UTC*
